Add piping for WebTransport statistics API calls

achristensen07 opened the pull request:
#### d21fd0303823030d06a1ec5e400a0d9b04afcb75
<pre>
Add piping for WebTransport statistics API calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=299935">https://bugs.webkit.org/show_bug.cgi?id=299935</a>
<a href="https://rdar.apple.com/160151501">rdar://160151501</a>

Reviewed by Matthew Finkel.

In order to have the context to make getStats calls, WebTransportReceiveStream
and WebTransportSendStream need to have a WebTransportSession and an identifier.
Rather than pass the identifier from WebKit to WebCore as separate objects,

I moved WebTransportSendStreamSink from WebKit to WebCore and gave it an identifier
accessor.  This makes WebTransportBidirectionalStreamConstructionParameters redundant,
but that can be simplified in a future PR.  This one&apos;s big enough.

The rest of the PR is just piping information around.  The stats are currently not
available, but we know we need to get them from the connection and connection group,
so this pipes the information to the right place and implements the simplest pieces
of it.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::createBidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::getStats):
(WebCore::WebTransport::createBidirectionalStream):
(WebCore::WebTransport::createUnidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransport.idl:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.cpp:
(WebCore::WebTransportBidirectionalStreamConstructionParameters::WebTransportBidirectionalStreamConstructionParameters):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.h:
* Source/WebCore/Modules/webtransport/WebTransportConnectionStats.h: Renamed from Source/WebCore/Modules/webtransport/WebTransportStats.h.
(WebCore::WebTransportConnectionStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportConnectionStats.idl: Renamed from Source/WebCore/Modules/webtransport/WebTransportStats.idl.
* Source/WebCore/Modules/webtransport/WebTransportDatagramStats.h:
(WebCore::WebTransportDatagramStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportDatagramStats.idl:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.cpp:
(WebCore::WebTransportReceiveStream::create):
(WebCore::WebTransportReceiveStream::WebTransportReceiveStream):
(WebCore::WebTransportReceiveStream::getStats):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamStats.h:
(WebCore::WebTransportReceiveStreamStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamStats.idl:
* Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp:
(WebCore::WebTransportSendStream::create):
(WebCore::WebTransportSendStream::WebTransportSendStream):
(WebCore::WebTransportSendStream::getStats):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.h:
* Source/WebCore/Modules/webtransport/WebTransportSendStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportSendStreamSink.cpp: Renamed from Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp.
(WebCore::WebTransportSendStreamSink::WebTransportSendStreamSink):
(WebCore::WebTransportSendStreamSink::~WebTransportSendStreamSink):
(WebCore::WebTransportSendStreamSink::write):
(WebCore::WebTransportSendStreamSink::close):
(WebCore::WebTransportSendStreamSink::error):
* Source/WebCore/Modules/webtransport/WebTransportSendStreamSink.h: Renamed from Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.h.
(WebCore::WebTransportSendStreamSink::create):
(WebCore::WebTransportSendStreamSink::identifier const):
* Source/WebCore/Modules/webtransport/WebTransportSendStreamStats.h:
(WebCore::WebTransportSendStreamStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportSendStreamStats.idl:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
(WebCore::WorkerWebTransportSession::streamSendBytes):
(WebCore::WorkerWebTransportSession::getStats):
(WebCore::WorkerWebTransportSession::getSendStreamStats):
(WebCore::WorkerWebTransportSession::getReceiveStreamStats):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::cancelReceiveStream):
(WebKit::NetworkTransportSession::cancelSendStream):
(WebKit::NetworkTransportSession::destroyStream):
(WebKit::NetworkTransportSession::getSendStreamStats):
(WebKit::NetworkTransportSession::getReceiveStreamStats):
(WebKit::NetworkTransportSession::getStats):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.cpp:
(WebKit::NetworkTransportStream::getSendStreamStats):
(WebKit::NetworkTransportStream::getReceiveStreamStats):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
* Source/WebKit/NetworkProcess/webtransport/WebTransport.serialization.in: Copied from Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in.
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::getStats):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::receiveLoop):
(WebKit::NetworkTransportStream::getSendStreamStats):
(WebKit::NetworkTransportStream::getReceiveStreamStats):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
(WebKit::WebTransportSession::getStats):
(WebKit::WebTransportSession::getSendStreamStats):
(WebKit::WebTransportSession::getReceiveStreamStats):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, ClientBidirectional)):

Canonical link: <a href="https://commits.webkit.org/300850@main">https://commits.webkit.org/300850@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/987ae5ba60c20e11921e21421b33594d1297c7d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124078 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/43782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130907 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125955 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52380 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94386 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127032 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/35473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/111000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74977 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29164 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74391 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105210 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133579 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/51020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38873 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102854 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51397 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102665 "Passed tests") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/48022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/26275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19500 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/50875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56645 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/50338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/53685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/52012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
```diff
@@ -38,8 +38,10 @@ class WorkerWebTransportSession : public WebTransportSession, public WebTranspor
+    void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::ref(); }
+    void deref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::deref(); }
+    ThreadSafeWeakPtrControlBlock& controlBlock() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::controlBlock(); }
+    size_t weakRefCount() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::weakRefCount(); }
```
Can use `WTF_ABSTRACT_THREAD_SAFE_REF_COUNTED_AND_CAN_MAKE_WEAK_PTR_IMPL`?

sysrqb commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
```diff
@@ -93,20 +94,38 @@ WebTransportServer echoServer([](ConnectionGroup group) -> ConnectionTask {
     [webView loadHTMLString:html baseURL:[NSURL URLWithString:@"https://webkit.org/"]];
-    EXPECT_WK_STREQ([webView _test_waitForAlert], "successfully read abc");
+    const char* expected =
+        "successfully read abcd"
```
Where does the `d` come from? And do you know why we receive 1 more byte than we sent (next line)?

sysrqb reviewed (approved):
Overall, looks okay

achristensen07 committed:
Add piping for WebTransport statistics API calls
https://bugs.webkit.org/show_bug.cgi?id=299935
rdar://160151501

Reviewed by Matthew Finkel.

In order to have the context to make getStats calls, WebTransportReceiveStream
and WebTransportSendStream need to have a WebTransportSession and an identifier.
Rather than pass the identifier from WebKit to WebCore as separate objects,

I moved WebTransportSendStreamSink from WebKit to WebCore and gave it an identifier
accessor.  This makes WebTransportBidirectionalStreamConstructionParameters redundant,
but that can be simplified in a future PR.  This one's big enough.

The rest of the PR is just piping information around.  The stats are currently not
available, but we know we need to get them from the connection and connection group,
so this pipes the information to the right place and implements the simplest pieces
of it.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::createBidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
(WebCore::WebTransport::getStats):
(WebCore::WebTransport::createBidirectionalStream):
(WebCore::WebTransport::createUnidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransport.idl:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.cpp:
(WebCore::WebTransportBidirectionalStreamConstructionParameters::WebTransportBidirectionalStreamConstructionParameters):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamConstructionParameters.h:
* Source/WebCore/Modules/webtransport/WebTransportConnectionStats.h: Renamed from Source/WebCore/Modules/webtransport/WebTransportStats.h.
(WebCore::WebTransportConnectionStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportConnectionStats.idl: Renamed from Source/WebCore/Modules/webtransport/WebTransportStats.idl.
* Source/WebCore/Modules/webtransport/WebTransportDatagramStats.h:
(WebCore::WebTransportDatagramStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportDatagramStats.idl:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.cpp:
(WebCore::WebTransportReceiveStream::create):
(WebCore::WebTransportReceiveStream::WebTransportReceiveStream):
(WebCore::WebTransportReceiveStream::getStats):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamStats.h:
(WebCore::WebTransportReceiveStreamStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamStats.idl:
* Source/WebCore/Modules/webtransport/WebTransportSendStream.cpp:
(WebCore::WebTransportSendStream::create):
(WebCore::WebTransportSendStream::WebTransportSendStream):
(WebCore::WebTransportSendStream::getStats):
* Source/WebCore/Modules/webtransport/WebTransportSendStream.h:
* Source/WebCore/Modules/webtransport/WebTransportSendStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportSendStreamSink.cpp: Renamed from Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp.
(WebCore::WebTransportSendStreamSink::WebTransportSendStreamSink):
(WebCore::WebTransportSendStreamSink::~WebTransportSendStreamSink):
(WebCore::WebTransportSendStreamSink::write):
(WebCore::WebTransportSendStreamSink::close):
(WebCore::WebTransportSendStreamSink::error):
* Source/WebCore/Modules/webtransport/WebTransportSendStreamSink.h: Renamed from Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.h.
(WebCore::WebTransportSendStreamSink::create):
(WebCore::WebTransportSendStreamSink::identifier const):
* Source/WebCore/Modules/webtransport/WebTransportSendStreamStats.h:
(WebCore::WebTransportSendStreamStats::isolatedCopy const):
* Source/WebCore/Modules/webtransport/WebTransportSendStreamStats.idl:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
(WebCore::WorkerWebTransportSession::streamSendBytes):
(WebCore::WorkerWebTransportSession::getStats):
(WebCore::WorkerWebTransportSession::getSendStreamStats):
(WebCore::WorkerWebTransportSession::getReceiveStreamStats):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::cancelReceiveStream):
(WebKit::NetworkTransportSession::cancelSendStream):
(WebKit::NetworkTransportSession::destroyStream):
(WebKit::NetworkTransportSession::getSendStreamStats):
(WebKit::NetworkTransportSession::getReceiveStreamStats):
(WebKit::NetworkTransportSession::getStats):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.cpp:
(WebKit::NetworkTransportStream::getSendStreamStats):
(WebKit::NetworkTransportStream::getReceiveStreamStats):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
* Source/WebKit/NetworkProcess/webtransport/WebTransport.serialization.in: Copied from Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in.
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::getStats):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::receiveLoop):
(WebKit::NetworkTransportStream::getSendStreamStats):
(WebKit::NetworkTransportStream::getReceiveStreamStats):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::createOutgoingUnidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
(WebKit::WebTransportSession::getStats):
(WebKit::WebTransportSession::getSendStreamStats):
(WebKit::WebTransportSession::getReceiveStreamStats):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, ClientBidirectional)):

Canonical link: https://commits.webkit.org/300850@main
